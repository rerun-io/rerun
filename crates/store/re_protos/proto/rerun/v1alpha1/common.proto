syntax = "proto3";

package rerun.common.v1alpha1;

// TODO(cmc): aren't most of the definitions in here dead code?

// supported encoder versions for encoding data
// See `RerunData` and `RerunChunkData` for its usage
enum EncoderVersion {
  ENCODER_VERSION_UNSPECIFIED = 0;
  ENCODER_VERSION_V0 = 1;
}

// RerunChunk is arrow IPC encoded RecordBatch that has
// rerun-specific semantic constraints and can be directly
// converted to a Rerun Chunk (`re_chunk::Chunk`)
message RerunChunk {
  // encoder version used to encode the data
  EncoderVersion encoder_version = 1;

  // Data payload is Arrow IPC encoded RecordBatch
  // TODO(zehiko) make this optional (#9285)
  bytes payload = 2;
}

// unique recording identifier. At this point in time it is the same id as the ChunkStore's StoreId
message RecordingId {
  string id = 1;
}

// uniquely identifies a table
message TableId {
  string id = 1;
}

// A recording can have multiple timelines, each is identified by a name, for example `log_tick`, `log_time`, etc.
message Timeline {
  string name = 1;
}

// A time range between start and end time points. Each 64 bit number can represent different time point data
// depending on the timeline it is associated with. Time range is inclusive for both start and end time points.
message TimeRange {
  int64 start = 1;
  int64 end = 2;
}

// arrow IPC serialized schema
message Schema {
  optional bytes arrow_schema = 1;
}

// TODO(cmc): can we kill this?
message Query {
  // The subset of the database that the query will run on: a set of EntityPath(s) and their
  // associated Component(s)
  ViewContents view_contents = 1;

  // Whether the view_contents should ignore semantically empty columns
  // A semantically empty column is a column that either contains no data at all, or where all
  // values are either nulls or empty arrays ([]).
  bool include_semantically_empty_columns = 2;

  // Whether the view_contents should ignore columns corresponding to indicator components
  // Indicator components are marker components, generally automatically inserted by Rerun, that
  // helps keep track of the original context in which a piece of data was logged/sent.
  bool include_indicator_columns = 3;

  // Whether the view_contents should ignore columns corresponding to Clear-related components
  bool include_tombstone_columns = 4;

  // The index used to filter out _rows_ from the view contents.
  // Only rows where at least 1 column contains non-null data at that index will be kept in the
  // final dataset. If left unspecified, the results will only contain static data.
  IndexColumnSelector filtered_index = 5;

  // The range of index values used to filter out _rows_ from the view contents
  // Only rows where at least 1 of the view-contents contains non-null data within that range will be kept in
  // the final dataset.
  // This has no effect if filtered_index isn't set.
  // This has no effect if using_index_values is set.
  IndexRange filtered_index_range = 6;

  // The specific index values used to filter out _rows_ from the view contents.
  // Only rows where at least 1 column contains non-null data at these specific values will be kept
  // in the final dataset.
  // This has no effect if filtered_index isn't set.
  // This has no effect if using_index_values is set.
  IndexValues filtered_index_values = 7;

  // The specific index values used to sample _rows_ from the view contents.
  // The final dataset will contain one row per sampled index value, regardless of whether data
  // existed for that index value in the view contents.
  // The semantics of the query are consistent with all other settings: the results will be
  // sorted on the filtered_index, and only contain unique index values.
  //
  // This has no effect if filtered_index isn't set.
  // If set, this overrides both filtered_index_range and filtered_index_values.
  IndexValues using_index_values = 8;

  // The component column used to filter out _rows_ from the view contents.
  // Only rows where this column contains non-null data be kept in the final dataset.
  ComponentColumnSelector filtered_is_not_null = 9;

  // The specific _columns_ to sample from the final view contents.
  // The order of the samples will be respected in the final result.
  //
  // If unspecified, it means - everything.
  ColumnSelection column_selection = 10;

  // Specifies how null values should be filled in the returned dataframe.
  SparseFillStrategy sparse_fill_strategy = 11;

  reserved 12;
  reserved "include_properties_entity";
}

message ColumnSelection {
  repeated ColumnSelector columns = 1;
}

message ColumnSelector {
  oneof selector_type {
    ComponentColumnSelector component_column = 2;
    TimeColumnSelector time_column = 3;
  }
}

message IndexColumnSelector {
  // TODO(zehiko) we need to add support for other types of index selectors
  Timeline timeline = 1;
}

message IndexRange {
  // TODO(zehiko) support for other ranges for other index selectors
  TimeRange time_range = 1;
}

message IndexValues {
  // TODO(zehiko) we need to add support for other types of index selectors
  repeated TimeInt time_points = 1;
}

message SampledIndexValues {
  repeated TimeInt sample_points = 1;
}

// A 64-bit number describing either nanoseconds, sequence numbers or fully static data.
message TimeInt {
  int64 time = 1;
}

message ViewContents {
  repeated ViewContentsPart contents = 1;
}

message ViewContentsPart {
  EntityPath path = 1;
  ComponentsSet components = 2;
}

message ComponentsSet {
  repeated Component components = 1;
}

// The unique identifier of an entity, e.g. `camera/3/points`
// See <https://www.rerun.io/docs/concepts/entity-path> for more on entity paths.
message EntityPath {
  string path = 1;
}

// Component describes semantic data that can be used by any number of  rerun's archetypes.
message Component {
  // component name needs to be a string as user can define their own component
  string name = 1;
}

// Used to telect a time column.
message TimeColumnSelector {
  Timeline timeline = 1;
}

// Used to select a component based on its EntityPath and Component name.
message ComponentColumnSelector {
  EntityPath entity_path = 1;
  Component component = 2;
}

// Specifies how null values should be filled in the returned dataframe.
enum SparseFillStrategy {
  SPARSE_FILL_STRATEGY_UNSPECIFIED = 0;
  SPARSE_FILL_STRATEGY_NONE = 1;
  SPARSE_FILL_STRATEGY_LATEST_AT_GLOBAL = 2;
}

message ApplicationId {
  string id = 1;
}

enum StoreKind {
  STORE_KIND_UNSPECIFIED = 0;
  STORE_KIND_RECORDING = 1;
  STORE_KIND_BLUEPRINT = 2;
}

message StoreId {
  StoreKind kind = 1;
  string id = 2;
}

message Tuid {
  // Approximate nanoseconds since epoch.
  optional fixed64 time_ns = 1;

  // Initialized to something random on each thread, then incremented for each
  // new `Tuid` being allocated.
  optional fixed64 inc = 2;
}

message EntryId {
  Tuid id = 1;
}

// Entry point for all ManifestRegistryService APIs
message DatasetHandle {
  // Unique entry identifier (for debug purposes)
  EntryId entry_id = 1;

  // Path to Dataset backing storage (e.g. s3://bucket/file or file:///path/to/file)
  optional string dataset_url = 2;
}

// DataframePart is arrow IPC encoded RecordBatch
message DataframePart {
  // encoder version used to encode the data
  EncoderVersion encoder_version = 1;

  // Data payload is Arrow IPC encoded RecordBatch
  optional bytes payload = 2;
}

// Generic parameters that will influence the behavior of the Lance scanner.
//
// TODO(zehiko, cmc): This should be available for every endpoint that queries data in
// one way or another.
message ScanParameters {
  // List of columns to project. If empty, all columns will be projected.
  repeated string columns = 1;

  IfMissingBehavior on_missing_columns = 2;

  // An arbitrary filter expression that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.filter(filter)
  // ```
  optional string filter = 3;

  // An arbitrary offset that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.limit(_, limit_offset)
  // ```
  optional int64 limit_offset = 4;

  // An arbitrary limit that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.limit(limit_len, _)
  // ```
  optional int64 limit_len = 5;

  // An arbitrary order clause that will be passed to the Lance scanner as-is.
  //
  // ```text
  // scanner.order_by(â€¦)
  // ```
  ScanParametersOrderClause order_by = 6;

  // If set, the output of `scanner.explain_plan` will be dumped to the server's log.
  bool explain_plan = 7;

  // If set, the final `scanner.filter` will be dumped to the server's log.
  bool explain_filter = 8;
}

enum IfMissingBehavior {
  IF_MISSING_BEHAVIOR_UNSPECIFIED = 0;
  IF_MISSING_BEHAVIOR_SKIP = 1;
  IF_MISSING_BEHAVIOR_ERROR = 2;
}

// Specify how the relevant creation call behaves
// in case of previously created (duplicate) items
enum IfDuplicateBehavior {
  IF_DUPLICATE_BEHAVIOR_UNSPECIFIED = 0;

  // Overwrite the existing item
  IF_DUPLICATE_BEHAVIOR_OVERWRITE = 1;

  // Skip if the item already exists
  IF_DUPLICATE_BEHAVIOR_SKIP = 2;

  // Return an error if the item already exists
  IF_DUPLICATE_BEHAVIOR_ERROR = 3;
}

message ScanParametersOrderClause {
  bool descending = 1;
  bool nulls_last = 2;
  optional string column_name = 3;
}

// Unique identifier for a partition. Can be user defined
// which means it can be of any type. For simplicity we start
// with a string, but we will probably revisit this.
message PartitionId {
  optional string id = 1;
}

message ComponentDescriptor {
  // Optional name of the `Archetype` associated with this data.
  optional string archetype_name = 1;

  // Optional name of the field within `Archetype` associated with this data.
  optional string archetype_field_name = 2;

  // Semantic name associated with this data.
  optional string component_name = 3;
}

// Unique identifier of a task submitted in the redap
// tasks subsystem
message TaskId {
  string id = 1;
}
