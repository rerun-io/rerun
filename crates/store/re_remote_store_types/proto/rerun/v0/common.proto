syntax = "proto3";

package rerun.remote_store.v0;

// unique recording identifier. At this point in time it is the same id as the ChunkStore's StoreId
message RecordingId {
    string id = 1;
}

// A recording can have multiple timelines, each is identified by a name, for example `log_tick`, `log_time`, etc.
message Timeline {
    string name = 1;
}

// A time range between start and end time points. Each 64 bit number can represent different time point data
// depending on the timeline it is associated with. Time range is inclusive for both start and end time points.
message TimeRange {
    int64 start = 1;
    int64 end = 2;
}

// arrow IPC serialized schema
message Schema {
    bytes arrow_schema = 1;
}

message Query {
    // The subset of the database that the query will run on: a set of EntityPath(s) and their
    // associated Component(s)
    ViewContents view_contents = 1;

    // Whether the view_contents should ignore semantically empty columns
    // A semantically empty column is a column that either contains no data at all, or where all
    // values are either nulls or empty arrays ([]).
    bool include_semantically_empty_columns = 2;

    // Whether the view_contents should ignore columns corresponding to indicator components
    // Indicator components are marker components, generally automatically inserted by Rerun, that
    // helps keep track of the original context in which a piece of data was logged/sent.
    bool include_indicator_columns = 3;

    // Whether the view_contents should ignore columns corresponding to Clear-related components
    bool include_tombstone_columns = 4;

    // The index used to filter out _rows_ from the view contents.
    // Only rows where at least 1 column contains non-null data at that index will be kept in the
    // final dataset. If left unspecified, the results will only contain static data.
    IndexColumnSelector filtered_index = 5;

    // The range of index values used to filter out _rows_ from the view contents
    // Only rows where at least 1 of the view-contents contains non-null data within that range will be kept in
    // the final dataset.
    // This has no effect if filtered_index isn't set.
    // This has no effect if using_index_values is set.
    IndexRange filtered_index_range = 6;

    // The specific index values used to filter out _rows_ from the view contents.
    // Only rows where at least 1 column contains non-null data at these specific values will be kept
    // in the final dataset.
    // This has no effect if filtered_index isn't set.
    // This has no effect if using_index_values is set.
    IndexValues filtered_index_values = 7;

    // The specific index values used to sample _rows_ from the view contents.
    // The final dataset will contain one row per sampled index value, regardless of whether data
    // existed for that index value in the view contents.
    // The semantics of the query are consistent with all other settings: the results will be
    // sorted on the filtered_index, and only contain unique index values.
    //
    // This has no effect if filtered_index isn't set.
    // If set, this overrides both filtered_index_range and filtered_index_values.
    IndexValues using_index_values = 8;

    // The component column used to filter out _rows_ from the view contents.
    // Only rows where this column contains non-null data be kept in the final dataset.
    ComponentColumnSelector filtered_is_not_null = 9;

    /// The specific _columns_ to sample from the final view contents.
    /// The order of the samples will be respected in the final result.
    ///
    /// If unspecified, it means - everything.
    ColumnSelection column_selection = 10;

    // Specifies how null values should be filled in the returned dataframe.
    SparseFillStrategy sparse_fill_strategy = 11;
}

message ColumnSelection {
    repeated ColumnSelector columns = 1;
}

message ColumnSelector {
    oneof selector_type {
        ComponentColumnSelector component_column = 2;
        TimeColumnSelector time_column = 3;
    }
}

message IndexColumnSelector {
    // TODO(zehiko) we need to add support for other types of index selectors
    Timeline timeline = 1;
}

message IndexRange {
    // TODO(zehiko) support for other ranges for other index selectors
    TimeRange time_range = 1;
}

message IndexValues {
    // TODO(zehiko) we need to add support for other types of index selectors
    repeated TimeInt time_points = 1;
}

message SampledIndexValues {
    repeated TimeInt sample_points = 1;
}

// A 64-bit number describing either nanoseconds, sequence numbers or fully static data.
message TimeInt {
    int64 time = 1;
}

message ViewContents {
    repeated ViewContentsPart contents = 1;
}

message ViewContentsPart {
    EntityPath path = 1;
    ComponentsSet components = 2;
}

message ComponentsSet {
    repeated Component components = 1;
}

// The unique identifier of an entity, e.g. `camera/3/points`
// See <https://www.rerun.io/docs/concepts/entity-path> for more on entity paths.
message EntityPath {
    string path = 1;
}

// Component describes semantic data that can be used by any number of  rerun's archetypes.
message Component {
    // component name needs to be a string as user can define their own component
    string name = 1;
}

// Used to telect a time column.
message TimeColumnSelector {
    Timeline timeline = 1;
}

// Used to select a component based on its EntityPath and Component name.
message ComponentColumnSelector {
    EntityPath entity_path = 1;
    Component component = 2;
}

// Specifies how null values should be filled in the returned dataframe.
enum SparseFillStrategy {
    NONE = 0;
    LATEST_AT_GLOBAL = 1;
}
