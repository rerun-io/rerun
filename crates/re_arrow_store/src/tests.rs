use arrow2::{
    array::{Array, ListArray},
    buffer::Buffer,
    chunk::Chunk,
};
use arrow2_convert::serialize::TryIntoArrow;
use polars::export::arrow::datatypes::Field;

use super::*;

/// Wrap `field_array` in a single-element `ListArray`
fn wrap_in_listarray(field_array: Box<dyn Array>) -> ListArray<i32> {
    ListArray::<i32>::from_data(
        ListArray::<i32>::default_datatype(field_array.data_type().clone()), // datatype
        Buffer::from(vec![0, field_array.len() as i32]),                     // offsets
        field_array,                                                         // values
        None,                                                                // validity
    )
}

/// Create `len` dummy rectangles
fn build_some_rects(len: usize) -> Box<dyn Array> {
    let v = (0..len)
        .into_iter()
        .map(|i| field_types::Rect2D {
            x: i as f32,
            y: i as f32,
            w: (i / 2) as f32,
            h: (i / 2) as f32,
        })
        .collect::<Vec<_>>();
    v.try_into_arrow().unwrap()
}

/// Create `len` dummy colors

fn build_some_colors(len: usize) -> Box<dyn Array> {
    let v = (0..len)
        .into_iter()
        .map(|i| i as field_types::ColorRGBA)
        .collect::<Vec<_>>();
    v.try_into_arrow().unwrap()
}

/// Create `len` dummy labels
fn build_some_labels(len: usize) -> Box<dyn Array> {
    let v = (0..len)
        .into_iter()
        .map(|i| format!("label{i}"))
        .collect::<Vec<_>>();
    v.try_into_arrow().unwrap()
}

/// Build a sample row of Rect data
fn build_test_rect_chunk() -> (Chunk<Box<dyn Array>>, ArrowSchema) {
    let time = arrow2::array::UInt32Array::from_slice([1234]).boxed();
    let rect = wrap_in_listarray(build_some_rects(5)).boxed();
    let color = wrap_in_listarray(build_some_colors(5)).boxed();
    let label = wrap_in_listarray(build_some_labels(1)).boxed();

    let schema = vec![
        Field::new("log_time", time.data_type().clone(), false),
        Field::new("rect", rect.data_type().clone(), true),
        Field::new("color", color.data_type().clone(), true),
        Field::new("label", label.data_type().clone(), true),
    ]
    .into();
    let chunk = Chunk::new(vec![time, rect, color, label]);
    (chunk, schema)
}

#[test]
fn test_rect_chunk() {
    let (chunk, schema) = build_test_rect_chunk();
    let x = polars::prelude::DataFrame::try_from((chunk, schema.fields.as_slice()));
    println!("{x:?}");
}

//--- Old tests --

#[test]
fn test_time_query() {
    let mut df1: DataFrame = df!(
        "time" => &[1, 3, 2],
        "numeric" => &[None, None, Some(3)],
        "object" => &[None, Some("b"), None],
        "dat" => &[Some(99), None, Some(66)],
    )
    .unwrap();

    let _df_sorted = df1.sort_in_place(["time"], false).unwrap();
}

#[test]
fn test_append_unified() {
    let mut df1 = df!(
        "colA" => [1, 2, 3],
        "colB" => ["one", "two", "three"],
    )
    .unwrap();

    let df2 = df!(
        "colA" => [4, 5, 6],
        "colC" => [Some(0.0), Some(0.1), None],
    )
    .unwrap();

    append_unified(&mut df1, &df2).unwrap();
}
