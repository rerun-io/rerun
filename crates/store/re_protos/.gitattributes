src/v0/** linguist-generated=true
