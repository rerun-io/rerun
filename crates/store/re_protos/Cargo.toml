[package]
name = "re_protos"
authors.workspace = true
description = "Rerun remote gRPC/protobuf API types"
edition.workspace = true
homepage.workspace = true
include.workspace = true
license.workspace = true
publish = true
readme = "README.md"
repository.workspace = true
rust-version.workspace = true
version.workspace = true

[dependencies]
re_log_types.workspace = true
re_dataframe.workspace = true

# External
arrow2 = { workspace = true, features = ["io_ipc"] }
prost.workspace = true
thiserror.workspace = true

# Native dependencies:
[target.'cfg(not(target_arch = "wasm32"))'.dependencies]
tonic = { workspace = true, default-features = false, features = [
  "codegen",
  "prost",
  "transport",
] }

# Web dependencies:
[target.'cfg(target_arch = "wasm32")'.dependencies]
tonic = { workspace = true, default-features = false, features = [
  "codegen",
  "prost",
] }
tonic-web-wasm-client.workspace = true

[lints]
workspace = true
