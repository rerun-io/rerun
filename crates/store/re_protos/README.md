# re_protos

Part of the [`rerun`](https://github.com/rerun-io/rerun) family of crates.

[![Latest version](https://img.shields.io/crates/v/re_protos.svg)](https://crates.io/crates/re_protos?speculative-link)
[![Documentation](https://docs.rs/re_protos/badge.svg)](https://docs.rs/re_protos?speculative-link)
![MIT](https://img.shields.io/badge/license-MIT-blue.svg)
![Apache](https://img.shields.io/badge/license-Apache-blue.svg)

Rerun remote store node gRPC API service types (client and server).

This crate includes both the language-agnostic definitions (protobuf) as well as the generated code.

The code is generated with `pixi run codegen-rstore`.
