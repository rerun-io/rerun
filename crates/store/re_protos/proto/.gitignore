schema_snapshot.main.yaml
