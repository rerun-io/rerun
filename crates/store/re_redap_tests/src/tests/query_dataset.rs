use futures::StreamExt as _;

use re_protos::{
    cloud::v1alpha1::{
        QueryDatasetResponse, ext::QueryDatasetRequest,
        rerun_cloud_service_server::RerunCloudService,
    },
    headers::RerunHeadersInjectorExt as _,
};

use crate::tests::common::{
    DataSourcesDefinition, LayerDefinition, RerunCloudServiceExt as _, concat_record_batches,
};
use crate::{FieldsExt as _, RecordBatchExt as _};

pub async fn query_empty_dataset(service: impl RerunCloudService) {
    let dataset_name = "dataset";
    service.create_dataset_entry_with_name(dataset_name).await;

    query_dataset_snapshot(
        &service,
        QueryDatasetRequest::default(),
        dataset_name,
        "empty_dataset",
    )
    .await;
}

pub async fn query_simple_dataset(service: impl RerunCloudService) {
    let data_sources_def = DataSourcesDefinition::new([
        LayerDefinition::simple("my_partition_id1", &["my/entity", "my/other/entity"]),
        LayerDefinition::simple("my_partition_id2", &["my/entity"]),
        LayerDefinition::simple(
            "my_partition_id3",
            &["my/entity", "another/one", "yet/another/one"],
        ),
    ]);

    let dataset_name = "dataset";
    service.create_dataset_entry_with_name(dataset_name).await;
    service
        .register_with_dataset_name(dataset_name, data_sources_def.to_data_sources())
        .await;

    let requests = vec![
        (QueryDatasetRequest::default(), "default"),
        (
            QueryDatasetRequest {
                partition_ids: vec!["my_partition_id3".into()],
                ..Default::default()
            },
            "single_partition",
        ),
        (
            QueryDatasetRequest {
                entity_paths: vec!["/my/entity".into()],
                select_all_entity_paths: false,
                ..Default::default()
            },
            "single_entity",
        ),
        //TODO(RR-2613): add more test cases here when they are supported by OSS server
    ];

    for (request, snapshot_name) in requests {
        query_dataset_snapshot(
            &service,
            request,
            dataset_name,
            &format!("simple_dataset_{snapshot_name}"),
        )
        .await;
    }
}

pub async fn query_simple_dataset_with_layers(service: impl RerunCloudService) {
    let data_sources_def = DataSourcesDefinition::new([
        LayerDefinition::simple("partition1", &["my/entity"]),
        LayerDefinition::simple("partition1", &["extra/entity"]).layer_name("extra"),
        LayerDefinition::simple("partition2", &["another/one"]).layer_name("base"),
        LayerDefinition::simple("partition2", &["extra/entity"]).layer_name("extra"),
        LayerDefinition::simple("partition3", &["i/am/alone"]),
    ]);

    let dataset_name = "dataset_with_layers";
    service.create_dataset_entry_with_name(dataset_name).await;
    service
        .register_with_dataset_name(dataset_name, data_sources_def.to_data_sources())
        .await;

    query_dataset_snapshot(
        &service,
        QueryDatasetRequest::default(),
        dataset_name,
        "simple_with_layer",
    )
    .await;
}

/// Test that failure cases return the correct error code.
pub async fn query_dataset_should_fail(service: impl RerunCloudService) {
    let dataset_name = "dataset";
    service.create_dataset_entry_with_name(dataset_name).await;

    let test_cases = vec![
        (
            "cannot specify entity paths if `select_all_entity_paths` is true",
            QueryDatasetRequest {
                entity_paths: vec!["/entity/path".into()],
                select_all_entity_paths: true,
                ..Default::default()
            },
            tonic::Code::InvalidArgument,
        ),
        //TODO(#11591): add more failure cases
    ];

    for (descr, request, expected_code) in test_cases {
        let response = service
            .query_dataset(tonic::Request::new(request.into()))
            .await;

        match response {
            Ok(_) => {
                panic!("expected failure with code {expected_code}, but got success ({descr})",);
            }
            Err(err) => {
                assert_eq!(
                    err.code(),
                    expected_code,
                    "expected failure with code {expected_code}, but got {err} ({descr})"
                );
            }
        }
    }
}

// ---

async fn query_dataset_snapshot(
    service: &impl RerunCloudService,
    query_dataset_request: QueryDatasetRequest,
    dataset_name: &str,
    snapshot_name: &str,
) {
    let chunk_info = service
        .query_dataset(
            tonic::Request::new(query_dataset_request.into())
                .with_entry_name(dataset_name)
                .unwrap(),
        )
        .await
        .unwrap()
        .into_inner()
        .flat_map(|resp| futures::stream::iter(resp.unwrap().data))
        .map(|dfp| dfp.try_into().unwrap())
        .collect::<Vec<_>>()
        .await;

    let merged_chunk_info = concat_record_batches(&chunk_info);

    // these are the only columns guaranteed to be returned by `query_dataset`
    let required_field = QueryDatasetResponse::fields();

    assert!(
        merged_chunk_info
            .schema()
            .fields()
            .contains_unordered(&required_field),
        "query dataset must return all guaranteed fields\nExpected: {:#?}\nGot: {:#?}",
        merged_chunk_info.schema().fields(),
        required_field,
    );

    let required_column_names = required_field
        .iter()
        .map(|f| f.name().as_str())
        .collect::<Vec<_>>();
    let required_chunk_info = merged_chunk_info.filtered_columns(&required_column_names);

    insta::assert_snapshot!(
        format!("{snapshot_name}_schema"),
        required_chunk_info.format_schema_snapshot()
    );

    // these columns are not stable, so we cannot snapshot them
    let filtered_chunk_info = required_chunk_info
        .unfiltered_columns(&[QueryDatasetResponse::FIELD_CHUNK_KEY])
        .auto_sort_rows()
        .unwrap();

    insta::assert_snapshot!(
        format!("{snapshot_name}_data"),
        filtered_chunk_info.format_snapshot(false)
    );
}
