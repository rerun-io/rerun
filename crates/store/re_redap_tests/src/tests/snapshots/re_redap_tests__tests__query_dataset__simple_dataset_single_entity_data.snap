---
source: crates/store/re_redap_tests/src/tests/query_dataset.rs
expression: filtered_chunk_info.format_snapshot(false)
---
┌──────────────────────────────────┬────────────────────┬───────────────────────┬───────────────────┬─────────────────┐
│ chunk_id                         ┆ chunk_partition_id ┆ rerun_partition_layer ┆ chunk_entity_path ┆ chunk_is_static │
╞══════════════════════════════════╪════════════════════╪═══════════════════════╪═══════════════════╪═════════════════╡
│ 00000000000000010000000000000001 ┆ my_partition_id1   ┆ base                  ┆ /my/entity        ┆ false           │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 00000000000000010000000000000002 ┆ my_partition_id1   ┆ base                  ┆ /my/entity        ┆ true            │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 00000000000000020000000000000001 ┆ my_partition_id2   ┆ base                  ┆ /my/entity        ┆ false           │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 00000000000000020000000000000002 ┆ my_partition_id2   ┆ base                  ┆ /my/entity        ┆ true            │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 00000000000000030000000000000001 ┆ my_partition_id3   ┆ base                  ┆ /my/entity        ┆ false           │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ 00000000000000030000000000000002 ┆ my_partition_id3   ┆ base                  ┆ /my/entity        ┆ true            │
└──────────────────────────────────┴────────────────────┴───────────────────────┴───────────────────┴─────────────────┘
